12Oleg/Node.js_Practice
# HW3 v1

**Assignment**
Expanding a Node.js Application with TypeScript and MongoDB Database

Objective
Extend an existing Node.js application, converted to TypeScript, which uses Express.js, Swagger documentation, and runs via Nodemon. Update its API structure to include "Movies" and "Genres," and implement the ability to search for movies by genre using a MongoDB database.

Requirements
    1. Connect to MongoDB Database:
        1. Set up MongoDB on your computer or use a cloud-based MongoDB service.
        2. Utilize the Mongoose library to establish a connection to the MongoDB database in your application.
        
    2. Creating an API for "Movies":
        1. Create API endpoints for "Movies" with CRUD operations (Create, Read, Update, Delete).
        2. Each movie should have the following fields:
            ▪ Title (String, required field)
            ▪ Description (String, required field)
            ▪ Release date (Date, required field)
            ▪ Genre (Array of Strings, required field)

    3. Creating an API for "Genres":
        1. Create API endpoints for "Genres" with CRUD operations (Create, Read, Update, Delete).
        2. Each genre should have the following field:
            ▪ Name (String, required field)

    4. Searching for Movies by Genre:
        1. Add a new API endpoint for searching movies by genre (e.g., /movies/genre/:genreName).
        2. Accept the genre as a request parameter and return a list of movies with the corresponding genre.
        
    5. Data Validation and Error Handling:
        1. Implement data validation for all incoming requests and error handling in your API.
        2. Ensure appropriate HTTP status codes and error messages.
        
    6. Updated Swagger Documentation:
        1. Expand the Swagger documentation to include the new APIs for "Movies" and "Genres," specifying available operations and providing examples of requests.
        
    7. Testing:
        1. Ensure that your application still functions correctly after updating the API structure, adding the MongoDB database, and enabling movie searches by genre.
        2. Use Nodemon for development and real-time testing.
